src/v/pandaproxy/schema_registry/test/test_json_schema.cc

```diff
@@ -124,3 +124,179 @@ SEASTAR_THREAD_TEST_CASE(test_make_valid_json_schema) {
+          "reader: {}, writer: {}", data.reader_schema, data.writer_schema))
+        try {
+            BOOST_CHECK_MESSAGE(
+              !pps::check_compatible(
+                pps::make_json_schema_definition(
+                  f.store,
+                  pps::make_canonical_json_schema(
+                    f.store,
+                    {pps::subject{"test"},
+                     {data.reader_schema, pps::schema_type::json}})
+                    .get())
+                  .get(),
+                pps::make_json_schema_definition(
+                  f.store,
+                  pps::make_canonical_json_schema(
+                    f.store,
+                    {pps::subject{"test"},
+                     {data.writer_schema, pps::schema_type::json}})
+                    .get())
+                  .get()),
+              "should not be compatible");
+        } catch (...) {
+            BOOST_CHECK_MESSAGE(
+              false,
+              fmt::format(
+                "terminated with exception {}", std::current_exception()));
+        }
+    };
+}
+
+struct compatible_test_case {
+    std::string_view reader_schema;
+    std::string_view writer_schema;
+};
+
+static constexpr auto compatible_test_cases
+  = std::to_array<compatible_test_case>({
+    // same type
+    {
+      .reader_schema = R"({"type": "boolean"})",
+      .writer_schema = R"({"type": "boolean"})",
+    },
+    {
+      .reader_schema = R"({"type": "null"})",
+      .writer_schema = R"({"type": "null"})",
+    },
+    // restrict types
+    {
+      .reader_schema = R"({"type": ["null", "boolean"]})",
+      .writer_schema = R"({"type": "null"})",
+    },
+    {
+      .reader_schema = R"({"type": ["number", "boolean"], "minimum": 10.2})",
+      .writer_schema = R"({"type": ["integer", "boolean"], "minimum": 11})",
+    },
+    // numeric checks
+    {
+      .reader_schema = R"({"type": "number"})",
+      .writer_schema
+      = R"({"type": "number", "minimum": 11, "exclusiveMinimum": true})",
+    },
+    {
+      .reader_schema = R"({"type": "number"})",
+      .writer_schema = R"({"type": "number", "maximum": 11})",
+    },
+    {
+      .reader_schema = R"({"type": "number", "minumum": 1.1, "maximum": 4})",
+      .writer_schema = R"({"type": "number", "minimum": 1.1, "maximum": 3.2})",
+    },
+    {
+      .reader_schema = R"({"type": "number", "multipleOf": 10})",
+      .writer_schema = R"({"type": "number", "multipleOf": 20})",
+    },
+    // string checks
+    {
+      .reader_schema = R"({"type": "string"})",
+      .writer_schema = R"({"type": "string", "minLength": 1, "maxLength": 10})",
+    },
+    {
+      .reader_schema = R"({"type": "string", "minLength": 1})",
+      .writer_schema = R"({"type": "string", "minLength": 2})",
+    },
+    {
+      .reader_schema = R"({"type": "string", "pattern": "^test"})",
+      .writer_schema = R"({"type": "string", "pattern": "^test"})",
+    },
+  });
+SEASTAR_THREAD_TEST_CASE(test_compatible_schemas) {
```

Comment:
nit: This looks like a copy of `test_incompatible_test_cases` with a boolean value negated. I'd probably merge the two tests to reduce duplication. You could even reuse the test case struct types.

Reply:
caught red handed